### PR TITLE
Remove mutual_rooms `update_user_directory` check, and add extra documentation

### PR DESCRIPTION
**Pull request by ShadowJonathan**
Resolves https://github.com/matrix-org/synapse/issues/10339

### Pull Request Checklist

* [x] Pull request is based on the develop branch
* [x] Pull request includes a [changelog file](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#changelog). The entry should:
  - Be a short description of your change which makes sense to users. "Fixed a bug that prevented receiving messages from other servers." instead of "Moved X method from `EventStore` to `EventWorkerStore`.".
  - Use markdown where necessary, mostly for `code blocks`.
  - End with either a period (.) or an exclamation mark (!).
  - Start with a capital letter.
  - Feel free to credit yourself, by adding a sentence "Contributed by @github_username." or "Contributed by [Your Name]." to the end of the entry.
* [x] Pull request includes a [sign off](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#sign-off)
* [x] [Code style](https://matrix-org.github.io/synapse/latest/code_style.html) is correct
  (run the [linters](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#run-the-linters))
  
`Signed-off-by: Jonathan de Jong <jonathan@automatia.nl>`

### CODE REVIEW

File: synapse/rest/client/shared_rooms.py [hunk 1]
```diff
@@ -48,13 +48,6 @@ async def on_GET(
         self, request: SynapseRequest, user_id: str
     ) -> Tuple[int, JsonDict]:
 
-        if not self.user_directory_active:
```
Comment (erikjohnston) [inline synapse/rest/client/shared_rooms.py:51]: This check probably wants to be replaced with `self.hs.config.userdirectory.user_directory_search_enabled` as we do for `/user_directory/search`:

https://github.com/matrix-org/synapse/blob/9e83521af860cb33a7459dbe74188ce5ef39f446/synapse/rest/client/user_directory.py#L61-L62

File: docs/workers.md [hunk 2]
```diff
@@ -467,10 +467,19 @@ the following regular expressions:
...
 jobs related to updating the user directory.
 
+Above endpoint is not *required* to be routed to this worker. By default,
+`update_user_directory` is set to `true`, which means the main process
+will handle updates. All workers configured with `client` can handle the above
+endpoint as long as either this worker or the main process are configured to
+handle it, and are online.
+
+If `update_user_directory` is set to `false`, and this worker is not running,
+the above endpoint may give outdated results.
```
Comment (erikjohnston) [inline docs/workers.md:481]: It looks like it will actually return nothing.

File: docs/workers.md [hunk 3]
```diff
@@ -467,10 +467,19 @@ the following regular expressions:
...
-When using this worker you must also set `update_user_directory: False` in the
+When using this worker you must also set `update_user_directory: false` in the
 shared configuration file to stop the main synapse running background
 jobs related to updating the user directory.
 
+Above endpoint is not *required* to be routed to this worker. By default,
+`update_user_directory` is set to `true`, which means the main process
+will handle updates. All workers configured with `client` can handle the above
+endpoint as long as either this worker or the main process are configured to
+handle it, and are online.
```
Comment (erikjohnston) [inline docs/workers.md:478]: Bleurgh, this is a bit annoying to document. Ideally we'd deprecate `synapse.app.user_dir` and replace it with an option in `generic_worker` (like for background tasks).

File: synapse/rest/client/shared_rooms.py (hunk 1 above)
Comment (ShadowJonathan) [reply synapse/rest/client/shared_rooms.py:51]: No, please review this comment; https://github.com/matrix-org/synapse/issues/10339#issuecomment-1039175679

TLDR; This PR explicitly deals with the removal of this line, the only thing this bit of code needs is for the user directory updater to exist *somewhere* in the cluster, and by default it'll exist on the main process.

The pros and cons of this have already been agreed upon by anoa and rich in that issue.

File: docs/workers.md (hunk 2 above)
Comment (ShadowJonathan) [reply docs/workers.md:481]: If the user directory worker has been enabled some time in the past, but it's been disabled some time thereafter, then the database might have stale information, so this might return outdated data.

File: synapse/rest/client/shared_rooms.py (hunk 1 above)
Comment (erikjohnston) [reply synapse/rest/client/shared_rooms.py:51]: > TLDR; This PR explicitly deals with the removal of this line, the only thing this bit of code needs is for the user directory updater to exist _somewhere_ in the cluster, and by default it'll exist on the main process.

if `self.hs.config.userdirectory.user_directory_search_enabled` is `False` then the user directory updater doesn't exist anywhere, as its completely disabled. The issue seems to be only discussing the worker case, rather than the user directory being disabled entirely? I don't really understand why we would want to process the request if the directory is disabled, we should either `400` or early return with an empty response like we do for the search endpoint.

File: docs/workers.md (hunk 2 above)
Comment (erikjohnston) [reply docs/workers.md:481]: Oh, right yup, sorry I misread which config option this was talking about.

File: docs/workers.md (hunk 3 above)
Comment (ShadowJonathan) [reply docs/workers.md:478]: There's some talk about this in #10339, I don't particularly know how to resolve this now.

File: synapse/rest/client/shared_rooms.py (hunk 1 above)
Comment (erikjohnston) [reply synapse/rest/client/shared_rooms.py:51]: I'm happy with 400ing and changing if the MSC decides something different :+1:

File: docs/workers.md (hunk 3 above)
Comment (anoadragon453) [reply docs/workers.md:478]: Given the scale of the discussion in #10339, I would document the current state now and get this piece of the puzzle merged. Then we can worry about from the transition from worker type -> generic worker option (https://github.com/matrix-org/synapse/issues/10441).

Comment (ShadowJonathan) [reply docs/workers.md:478]: So should I additionally document the `synapse.app.user_dir` worker type in `workers.md`?

Comment (ShadowJonathan) [reply docs/workers.md:478]: Actually, the worker type is already documented, i think that we shouldnt encode future intentions in the documentation, until we actually are working on it.

File: synapse/rest/client/shared_rooms.py [hunk 4]
```diff
@@ -43,16 +43,17 @@ def __init__(self, hs: "HomeServer"):
...
         self.store = hs.get_datastores().main
         self.user_directory_active = hs.config.server.update_user_directory
+        self.hs = hs
 
     async def on_GET(
         self, request: SynapseRequest, user_id: str
     ) -> Tuple[int, JsonDict]:
 
-        if not self.user_directory_active:
+        if not self.hs.config.userdirectory.user_directory_search_enabled:
```
Comment (erikjohnston) [inline synapse/rest/client/shared_rooms.py:52]: Can we make this a variable rather than saving `self.hs`? It's a bit of a smell to have `self.hs` lying around if you can avoid it.

**Review by erikjohnston — COMMENTED** [review]
Otherwise LGTM

**Comment by ShadowJonathan** [comment]
I'll rebase this on https://github.com/matrix-org/synapse/pull/12036

**Review by erikjohnston — APPROVED** [review]
